setup tools

DanielDecker opened the pull request:
This PR introduces install scripts for the first time setup and on demand dependency installation discussed in #116.

* The `setup.sh` guides the user through what used to be a step-by-step guide in the readme. It installs the base dependencies via install_dependencies.sh, creates the python virtual env and optionally installs the service file and optional dependencies.
* The `install_dependencies.sh` installs base and optional plug-in dependencies listed in the `dependencies.txt` files, located in the main and plug-in directories. It is called from setup.sh but can also be called manually. If called e. g. with the argument 'bt' all dependencies listed in 'bt/dependencies.txt' get installed. I think this is a good compromise.

I have tested both scripts with Raspberry Pi OS bookworm and bullseye. The applications used in the scripts are pre-installed on both versions.

The requirements.txt has been removed as it's no longer needed.

DanielDecker commented:
@rkoshak : this RP is ready for merge.

rkoshak commented:
I've only just started a review and then we had a broken pipe at the house and I've not had a chance to get back to it. 

It's high on my list but I've not yet completed a review. The big thing is I want it to work for both bookworm and Ubuntu at a minimum.

DanielDecker commented:
Sorry, I didn't mean to rush you. Take your time.

I didn't tested the setup script on Ubuntu. It should also work on Ubuntu, at worst some of the script's dependencies are not met by the default installation.

DanielDecker commented:
I fixed lint errors in the bash scripts shown by 'shellcheck' and tested the changes.
@rkoshak : this PR is ready for merge.

rkoshak commented on gpio/README.md:
````diff
@@ -390,13 +378,12 @@ A recieved command will be sent back on all configured connections to the config
+cd /srv/sensorReporter
+sudo ./install_dependencies.sh gpio
 ```
 
````
We should probably offer an "uninstall" option. At a minimum it should remove the .system file and logs folder.

rkoshak replied:
> Would that satisfy your request?

Yes, that would work for me.

> Should we mention this somewhere in the readme?
Absolutely the uninstall option for install_dependencies.sh should be mentioned. That's a good option to make clear is there. 

For me at least it's not uncommon for me to move stuff from one RPi to another which means changing the sensorReporter configs.

DanielDecker replied:
I'll update this PR with the changes.
At the moment I don't have the time. This should change by the end of February.

DanielDecker replied:
done

rkoshak reviewed (commented):
It seems to work on my system. I didn't try everything but the basics seem to work. I've just a couple of comments.

Sorry for the tardy review.

DanielDecker commented on setup.sh:
```diff
@@ -0,0 +1,125 @@
+	read -r INST_SERVICE
+else
+	# if setup runs from a different folder
+	echo "Warn: Setup expects working directory '$SETUP_FOLDER', $APP_NAME service cannot be installed for this path."
```
This warning is related to the sensor_reporter.service, where the path to sensor_reporter is defined.
The setup can run from every folder (inside SensorReporter base folder) and install all except the service correctly.

rkoshak replied:
Maybe the script can do a quick `sed` operation to replace the path in the .service file with the path where sensorReporter is being installed. I know this sort of thing is routinely done in setup scripts. I use Ansible so can't readily supply the command though. In Ansible I'd use a template.

DanielDecker replied:
I'll look into it

DanielDecker replied:
done

DanielDecker commented:
I added all the requested features.
Modifying the sensor_reporter.service file was quite a challenge, as systemd has a special way of escaping special characters and 'sed' resolves escaped characters. I ended up escaping the input twice to get the desired result.
`./setup.sh --uninstall` will not remove the Python virtual env, the service, the logging folder and the sensorReporter user. I also added a '--yes' option to run setup.sh without interaction.

While testing I had a runtime error in the mqtt connector which is related to paho-mqtt version 2.0 release. So I looked the paho-mqtt version to >=1.6.1,<2.0.0 (in the mqtt/dependencies.txt).

I tested all the changes and the PR is ready to merge.

rkoshak commented:
It took me a couple days but it looks good to me. Thanks a bunch!

DanielDecker commented:
With pleasure! :-)
